em++

```diff
@@ -26,4 +26,9 @@ if [ -z "$PYTHON" ]; then
   exit 1
 fi
 
-exec "$PYTHON" "$0.py" "$@"
+if [ -z "$EMCC_CCACHE" ]; then
+  exec "$PYTHON" "$0.py" "$@"
+else
+  unset EMCC_CCACHE
+  exec ccache "$0" "$@"
```

Comment:
Why do we need to handle injecting ccache like this when other compilers such as clang and gcc don't?  Can't we ask folks to use cache in the normal way (which I believe is to perpend it to compile command?)

Reply:
>  Can't we ask folks to use cache in the normal way

This is the/a normal way. Ccache supports two methods of installation: 1) explicit invocation with `ccache gcc ...`, and 2) automatic invocation with `gcc ...`. [See the documentation of ccache](https://github.com/ccache/ccache/blob/master/doc/INSTALL.md).

> (which I believe is to perpend it to compile command?)

That is not the full picture. See the docs above. There are two recommended ways to install ccache, this integration is ensuring that both usages work.

E.g. FreeBSD man documentation: https://www.freebsd.org/cgi/man.cgi?query=ccache&sektion=1&n=1

```
RUN MODES
       There are two ways to use ccache. You can either	prefix your
       compilation commands with ccache	or you can let ccache masquerade as
       the compiler by creating	a symbolic link	(named as the compiler)	to
       ccache. The first method	is most	convenient if you just want to try out
       ccache or wish to use it	for some specific projects. The	second method
       is most useful for when you wish	to use ccache for all your
       compilations.
```

> Why do we need to handle injecting ccache like this when other compilers such as clang and gcc don't?

The reason that symbolic links are not used by this integration is because that would intrude with emsdk installation of either ccache or emscripten packages (say, e.g. if installing ccache package would modify installed emscripten package)

Also while [Linux and macOS could use symbolic links to inject ccache](https://github.com/ccache/ccache/blob/master/doc/INSTALL.md),  that is not well supported on Windows.

Reply:
My thoughts about this in general: if neither gcc nor clang require explicit in-tree modification to make ccache work I would hope that we do not need this either.

If seems that we are adding a third `RUN MODE` here that is emscripten-specific:

1. Run explicitly via `ccache <compiler>`
2. Run implicitly via having a `<compiler> -> ccache` symlink in your PATH
3. Run implicitly via add `EMCC_CACHE` to the environment.

Given that mode (3) does not exist today with existing compilers I don't see why we in emscripten want to be special and add this third mode.   I understand that mode (2) is hard/impossible on windows but that is pre-existing condition unrelated to emscripten, right?

Reply:
I don't agree here. Given that we can offer an easy "emsdk activate ccache" feature that just does the full integration, I don't understand why we should not do it. It gives a straightforward "fully set up out of the box" experience without manual hassles. Option 2 is a complex burden on Windows, but also on non-Windows platforms.

Also, the step (3) does not need to be something public, but an internal implementation aid to carry the integration together. I.e. we are not offering a "third run mode" here.

Reply:
Ok, I get that we want this to work out-of-the-box for all emsdk users including windows users.

I'd still like find a solution that doesn't involved modifying emscripten.   How about this solution:

1. create `emsdk/ccache/bin/emcc` and `emsdk/ccache/bin/em++`  (these are bat/sh wrappers rather than symlinks)
2. Have `emsdk activate ccache` put `emsdk/ccache/bin` first in the PATH before `emscripten/bin`.

This avoids changing emscripten at all and works more in like the symlink approach that exists today for ccache (but also happens to work on windows).    It also works with older/current versions of emscripten too.

This means figuring out how to chain `emsdk/ccache/bin/emcc` -> `emscripten/bin/emcc` in the wrapper script.   I don't know of the top of my head how to do that.    If its too tricky I'm not going to resist this emscripten-side change too much,  but it does seem wrong to me that we are hacking the compiler to be specifically aware of ccache (unlike gcc or clang).